code/modules/power/batteryrack.dm

```diff
@@ -1,234 +1,313 @@
+	input_level = max_transfer_rate
+	output_level = max_transfer_rate
+
+/obj/machinery/power/smes/batteryrack/Destroy()
+	for(var/obj/item/weapon/cell/C in internal_cells)
+		qdel(C)
+	internal_cells = null
+	return ..()
 
 /obj/machinery/power/smes/batteryrack/update_icon()
 	overlays.Cut()
-	if(stat & BROKEN)	return
-
-	if(!br_cache)
-		br_cache = list()
-		br_cache.len = 7
-		br_cache[1] = image('icons/obj/power.dmi', "gsmes_outputting")
-		br_cache[2] = image('icons/obj/power.dmi', "gsmes_charging")
-		br_cache[3] = image('icons/obj/power.dmi', "gsmes_overcharge")
-		br_cache[4] = image('icons/obj/power.dmi', "gsmes_og1")
-		br_cache[5] = image('icons/obj/power.dmi', "gsmes_og2")
-		br_cache[6] = image('icons/obj/power.dmi', "gsmes_og3")
-		br_cache[7] = image('icons/obj/power.dmi', "gsmes_og4")
-
-	if (output_attempt)
-		overlays += br_cache[1]
-	if(inputting)
-		overlays += br_cache[2]
-
-	var/clevel = chargedisplay()
-	if(clevel>0)
-		overlays += br_cache[3+clevel]
-	return
+	icon_update = 0
+
+	var/cellcount = 0
+	var/charge_level = between(0, round(Percentage() / 12), 7)
+
+
+	overlays += "charge[charge_level]"
+
+	for(var/obj/item/weapon/cell/C in internal_cells)
+		cellcount++
+		overlays += "cell[cellcount]"
+		if(C.fully_charged())
+			overlays += "cell[cellcount]f"
+		else if(!C.charge)
+			overlays += "cell[cellcount]e"
+
+// Recalculate maxcharge and similar variables.
+/obj/machinery/power/smes/batteryrack/proc/update_maxcharge()
+	var/newmaxcharge = 0
+	for(var/obj/item/weapon/cell/C in internal_cells)
+		newmaxcharge += C.maxcharge
+
+	newmaxcharge /= CELLRATE		// Convert to Joules
+	newmaxcharge *= SMESRATE		// And to SMES charge units (which are for some reason different than CELLRATE)
+	capacity = newmaxcharge
+	charge = between(0, charge, newmaxcharge)
+
+
+// Sets input/output depending on our "mode" var.
+/obj/machinery/power/smes/batteryrack/proc/update_io(var/newmode)
+	mode = newmode
+	switch(mode)
+		if(PSU_OFFLINE)
+			input_attempt = 0
+			output_attempt = 0
+		if(PSU_INPUT)
+			input_attempt = 1
+			output_attempt = 0
+		if(PSU_OUTPUT)
+			input_attempt = 0
+			output_attempt = 1
+		if(PSU_AUTO)
+			input_attempt = 1
+			output_attempt = 1
+
+// Store charge in the power cells, instead of using the charge var. Amount is in joules.
+/obj/machinery/power/smes/batteryrack/add_charge(var/amount)
+	amount *= CELLRATE // Convert to CELLRATE first.
+	if(equalise)
+		// Now try to get least charged cell and use the power from it.
+		var/obj/item/weapon/cell/CL = get_least_charged_cell()
+		amount -= CL.give(amount)
+		if(!amount)
+			return
+	// We're still here, so it means the least charged cell was full OR we don't care about equalising the charge. Give power to other cells instead.
+	for(var/obj/item/weapon/cell/C in internal_cells)
+		amount -= C.give(amount)
+		// No more power to input so return.
+		if(!amount)
+			return
+
+
+/obj/machinery/power/smes/batteryrack/remove_charge(var/amount)
```

Comment:
Should this return some feedback on whether it was able to remove the desired amount? Or does it check that it has enough charge to remove before calling this?

Reply:
Double-checked this, it does check there is enough charge before this is called.